{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GWAS_IBD']\n",
      "Number of unique SNPs: 639\n",
      "all upstream genes: 166\n",
      "all downstream genes: 166\n",
      "all SNP genes: 279\n",
      "all genes: 552\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#\n",
    "# parsing GWAS for IBD paper\n",
    "#\n",
    "\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "filename='../../IBD_paper/genes/Supplementary Document 1.xlsx'\n",
    "xls = pd.ExcelFile(filename)\n",
    "print(xls.sheet_names)\n",
    "\n",
    "GWAS_table = pd.read_excel(xls, 'GWAS_IBD')\n",
    "#print(GWAS_table.head())\n",
    "\n",
    "##printing cols\n",
    "#for col in GWAS_table.columns: \n",
    "#    print(col) \n",
    "unique_SNPs= GWAS_table['SNPS'].unique()\n",
    "print( 'Number of unique SNPs: %s' %len(unique_SNPs))\n",
    "all_SNP_genes=list(GWAS_table['SNP_GENE_IDS'].unique())\n",
    "all_upstream_genes=list(GWAS_table['UPSTREAM_GENE_ID'].unique())\n",
    "all_downstream_genes=list(GWAS_table['DOWNSTREAM_GENE_ID'].unique())\n",
    "\n",
    "\n",
    "all_SNP_genes.remove('no')\n",
    "all_upstream_genes.remove('no')\n",
    "all_downstream_genes.remove('no')\n",
    "print('all upstream genes: %s' %len(all_upstream_genes))\n",
    "print('all downstream genes: %s' %len(all_downstream_genes))\n",
    "print('all SNP genes: %s' %len(all_SNP_genes))\n",
    "\n",
    "all_GWAS_genes=set(all_SNP_genes + all_upstream_genes + all_downstream_genes)\n",
    "\n",
    "print('all genes: %s' %len(all_GWAS_genes))\n",
    "#print(GWAS_table['SNPS'].unique())\n",
    "\n",
    "# make dictionary with keys \n",
    "d_GWAS_SNPs=dict.fromkeys(set(unique_SNPs), [])\n",
    "#d_genes=dict.fromkeys(set(all_GWAS_genes), [])\n",
    "#print(d_genes)\n",
    "#print(d_GWAS_SNPs)\n",
    "d_genes={}\n",
    "\n",
    "fp=open('../../IBD_paper/genes/Supplementary Document 1.txt','r', encoding='utf-8', errors='ignore')\n",
    "for line in fp:\n",
    "    if line[0]=='#':\n",
    "        continue\n",
    "    data      = line.strip().split('\\t')\n",
    "    try:\n",
    "        upstream    = int(data[15])\n",
    "        \n",
    "    except:\n",
    "        if ',' in data[15]:\n",
    "            upstream_genes=data[15].split(',')\n",
    "            for what in upstream_genes:\n",
    "                if what in d_genes.keys():\n",
    "                    d_genes[what]+=';'+SNP\n",
    "                else:\n",
    "                    d_genes[what]=SNP\n",
    "            \n",
    "    try:\n",
    "        downstream  = int(data[16])\n",
    "    except:\n",
    "        if ',' in data[16]:\n",
    "            downstream_genes=data[16].split(',')\n",
    "            for whit in downstream_genes:\n",
    "                if whit in d_genes.keys():\n",
    "                    d_genes[whit]+=';'+SNP\n",
    "                else:\n",
    "                    d_genes[whit]=SNP\n",
    "    try:\n",
    "        SNP_gene    = int(data[17])\n",
    "    except:\n",
    "        if ',' in data[17]:\n",
    "            SNP_genes=data[17].split(',')\n",
    "            for whot in SNP_genes:\n",
    "                if whot in d_genes.keys():\n",
    "                    d_genes[whot]+=';'+SNP\n",
    "                else:\n",
    "                    d_genes[whot]=SNP\n",
    "        \n",
    "    SNP = str(data[21])\n",
    "    \n",
    "    if upstream in d_genes.keys():\n",
    "        d_genes[upstream]+=';'+SNP\n",
    "    else:\n",
    "        d_genes[upstream]=SNP\n",
    "    \n",
    "    if downstream in d_genes.keys():\n",
    "        d_genes[downstream]+=';'+SNP\n",
    "    else:\n",
    "        d_genes[downstream]=SNP\n",
    "        \n",
    "    if SNP_gene in d_genes.keys():\n",
    "        d_genes[SNP_gene]+=';'+SNP\n",
    "    else:\n",
    "        d_genes[SNP_gene]=SNP\n",
    "fp.close()\n",
    "\n",
    "#print(d_genes)\n",
    "\n",
    "#\n",
    "# unique SNPs \n",
    "d_genes_unique={}\n",
    "for gene, SNPs in d_genes.items():\n",
    "    list_of_SNPs=SNPs.split(';')\n",
    "    set_of_SNPs=set(list_of_SNPs)\n",
    "    d_genes_unique[gene]=set_of_SNPs\n",
    "\n",
    "    \n",
    "    \n",
    "#print(d_genes_unique)\n",
    "f=open('../../IBD_paper/genes/all_GWAS_genes_with_snps.txt','w')\n",
    "f.write('gene\\tSNP\\n')\n",
    "for gene, SNPs in d_genes_unique.items(): \n",
    "    f.write('%s\\t' %gene)\n",
    "    for SNP in SNPs:\n",
    "        f.write('%s;' %SNP)\n",
    "    f.write('\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
